### PR TITLE
Build LLVM from sources

### LINKED ISSUE #66493: Compile clang/llvm from sources
There are multiple reasons to do this:
- unable to pin compiler to specific version, since apt.llvm.org contains only latest minor version for each major
- `--target` mismatch - if `--target` does not match `LLVM_DEFAULT_TARGET_TRIPLE` then clang will not find runtime sanitizers libraries (but right now there is a workaround in docker build images)
- _patching_

Refs:
- https://github.com/llvm/llvm-project/pull/97802
- https://github.com/ClickHouse/ClickHouse/issues/66049

Cc: @alesapin

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Pin the compiler and tools version by building LLVM from sources. This eliminates a fragile supply-chain dependency. This closes #66493. See https://github.com/llvm/llvm-project/issues/99453

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 476d9d930b4f4ed6929ca1ef717349d0ab702692 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/66793/476d9d930b4f4ed6929ca1ef717349d0ab702692/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/66778/6aa754decf9ce156c8e912e9f4e3a0b5e0ece9a8/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: docker/test/util/Dockerfile [hunk 1]
```diff
@@ -27,42 +20,40 @@ RUN apt-get update \
...
-    && rm -rf /var/lib/apt/lists/* /var/cache/debconf /tmp/*
+        git \
+        ninja-build \
+        clang \
+        --yes --no-install-recommends --verbose-versions \
+        && apt-get clean && rm -rf /var/lib/apt/lists/* /var/cache/debconf /tmp/*
+
+# Install cmake 3.20+ for Rust support
+# Used https://askubuntu.com/a/1157132 as reference
+RUN curl -s https://apt.kitware.com/keys/kitware-archive-latest.asc | gpg --dearmor - > /etc/apt/trusted.gpg.d/kitware.gpg \
```
Comment (thevar1able) [inline docker/test/util/Dockerfile:41]: We may merge this in one big `RUN` so that we don't have intermediate layers with unnecessary dependencies.

**Review by alesapin — CHANGES_REQUESTED** [review]
I'm strongly against building such complex tools as clang or llvm at docker build stage. It's very unclear, implicit, and slow down these jobs. All other CI jobs depend on them.

`docker build` command shouldn't take hours if your laptop is not powerful enough. 

 Let's introduce a separate job for it which will publish artifacts to S3. We can even introduce fancy URLs for them and it will be available for every clickhouse developer.

**Comment by thevar1able** [comment]
I somewhat agree with @alesapin, this approach will work, but is very inefficient and will slow down CI significantly. Would be much nicer to make something like `FROM clickhouse/llvm:<llvm-hash>`, then that image could be rebuilt only on LLVM commit hash update.

**Comment by alexey-milovidov** [comment]
The `base` Docker image is rebuild only when it's changed, so it isn't a problem.

**Comment by alesapin** [comment]
It's a huge problem. What if I want to use locally exactly the same compiler/llvm which CI uses? I have to rebuild it (what if my PC is slow), or execute weird commands to extract it from docker image. Separate job with artifacts is much better than this solution.

**Comment by alexey-milovidov** [comment]
Ok, but this PR is still a step forward. Today you have a difficult way to get exactly the same compiler, either on your machine or in CI. After this pull request, it becomes easier.

**Comment by alexey-milovidov** [comment]
Reopen if needed.

**Pull request closed without merging**